pkg/querier/multi_tenant_querier.go

```diff
@@ -84,17 +86,18 @@ func (q *MultiTenantQuerier) SelectSamples(ctx context.Context, params logql.Sel
 	}
 	params.Selector = updatedSelector.String()
 
-	iters := make([]iter.SampleIterator, len(matchedTenants))
-	i := 0
+	p := pool.NewWithResults[iter.SampleIterator]().WithContext(ctx).WithCancelOnError()
 	for id := range matchedTenants {
-		singleContext := user.InjectOrgID(ctx, id)
-		iter, err := q.Querier.SelectSamples(singleContext, params)
-		if err != nil {
-			return nil, err
-		}
-
-		iters[i] = NewTenantSampleIterator(iter, id)
-		i++
+		id := id
+		p.Go(func(c context.Context) (iter.SampleIterator, error) {
+			singleContext := user.InjectOrgID(c, id)
+			iter, err := q.Querier.SelectSamples(singleContext, params)
```

Comment:
This fails with a cancelled context when run a second time. I have a hard time understanding why.